### PR TITLE
Enable http2 support between proxy and member

### PR DESCRIPTION
**Pull request by ZhuPeng**
Issue https://github.com/coreos/etcd/issues/5033. It was just enable the http2 support to proxy and member, and it was enable in client connecting to the proxy. I thought it maybe support when using v3 api.

### CODE REVIEW

**Comment by xiang90** [comment]
@ZhuPeng Can you try to set the tls config to the proxy rather than calling this function explicitly?

**Comment by xiang90** [comment]
kindly ping

**Comment by ZhuPeng** [comment]
Sorry, I'm a little busy recently. We just copy the `http2.ConfigureTransport` implements to here? And I am confuse to put the tls config only in proxy or put it to the `NewTransport` https://github.com/coreos/etcd/blob/master/pkg/transport/listener.go#L60 .

**Comment by xiang90** [comment]
@ZhuPeng No. I believe all we need to do is to change https://github.com/coreos/etcd/blob/master/etcdmain/etcd.go#L520 to a http.Server and pass in tls configuration when needed.

**Comment by xiang90** [comment]
like what we do here: https://github.com/coreos/etcd/blob/master/etcdmain/serve.go#L67-L72

**Comment by ZhuPeng** [comment]
We can divide it to two separate problems. 
The first is enable HTTP2 support for proxy connneting to cluster members. If we use an transport with tls config to connecting to etcd cluster member, which support HTTP2, the transport will not use HTTP2 automaticly(issue #5033). 

And the second is enable HTTP2 support for proxy server, which serve the client request. This problem can solve by the solution we did in etcd server. But I thought it is not the very import thing, my requirements is deploy a great many of proxy and every proxy would serve very few client requests.

**Comment by xiang90** [comment]
> The first is enable HTTP2 support for proxy connneting to cluster members. If we use an transport with tls config to connecting to etcd cluster member, which support HTTP2, the transport will not use HTTP2 automaticly(issue #5033).

Have you tried to compile both proxy and member with go1.6? And http2 does not work automatically for the proxy <-> member part communication? From my pervious experience, it should just work.

**Comment by xiang90** [comment]
OK. Now I see what you are saying by looking through http2 code.

**Comment by xiang90** [comment]
@ZhuPeng Sorry about the noise I made. Can you please follow https://github.com/coreos/etcd/blob/master/CONTRIBUTING.md#format-of-the-commit-message to format your commit message? Also can you please add comment to the code? Probably just link to the go issue about why go does not configure transport by default for now. Thanks a lot.

**Comment by ZhuPeng** [comment]
Ok, no problem.

File: proxy/proxy.go [hunk 1]
```diff
@@ -42,6 +44,14 @@ type GetProxyURLs func() []string
 // which will proxy requests to an etcd cluster.
 // The handler will periodically update its view of the cluster.
 func NewHandler(t *http.Transport, urlsFunc GetProxyURLs, failureWait time.Duration, refreshInterval time.Duration) http.Handler {
+	if t.TLSClientConfig != nil {
+		// Enable http2, see Issue 5033.
```
Comment (xiang90) [inline proxy/proxy.go]: probably point to https://github.com/golang/go/issues/14275 instead?

File: proxy/proxy.go [hunk 2]
```diff
@@ -42,6 +44,14 @@ type GetProxyURLs func() []string
...
+		err := http2.ConfigureTransport(t)
+		if err != nil {
+			plog.Infof("Error enabling Transport HTTP/2 support: %v", err)
```
Comment (xiang90) [inline proxy/proxy.go]: this should be a warning?

**Comment by xiang90** [comment]
@ZhuPeng Can you rebase this with current master? I moved proxy to httpproxy pkg. Sorry.

**Comment by xiang90** [comment]
@ZhuPeng Thanks. Please ping me if you push a new commit. github is stupid about that... I do not have any notification if you do not ping me.

**Pull request merged**